Satellite graphs docs draft (#357)
arangodb/docs: 3.7/graphs-satellite-graphs.md

---
layout: default
description: Graphs synchronously replicated to all servers, available in the Enterprise Edition
title: ArangoDB Satellite Graphs
---
Satellite Graphs
=====================

{% hint 'info' %}
Satellite Graphs are only available in the
[**Enterprise Edition**](https://www.arangodb.com/why-arangodb/arangodb-enterprise/){:target="_blank"},
and the [**ArangoDB Cloud**](https://cloud.arangodb.com/){:target="_blank"}.
{% endhint %}

Satellite Graphs are the natural extension of the concept of Satellite
collections to graphs. All of the usual benefits and caveats apply.

When doing joins involving graph traversals, shortest paths, or k-shortest paths
in an ArangoDB cluster, data has to be exchanged between different servers. In particular
graph traversals are usually executed on a Coordinator, because they need global information. 

This results in a lot of network traffic and slow query execution.

Satellite graphs are synchronously replicated to all DB-Servers that are part of

> a cluster, which enables DB-Servers to execute parts of joins locally.

Reviewer: ```suggestion
a cluster, which enables DB-Servers to execute graph traversals (and (k-)shortest paths), and possibly joins with traversals, locally.
```

> This greatly improves performance for such joins.

Reviewer: ```suggestion
This greatly improves performance for such queries.
```

To create a satellite graph in arangosh, use the `satelliteGraph` module.

    arangosh> satGM = require("@arangodb/satellite-graph")
    arangosh> satGM._create("satelliteGraph", [ satGM._relation("edges", "vertices", "vertices") ], [], {});

> You can also create vertex and edge collections manually, and set their `replicationFactor` to

Reviewer 2: I think I would remove that section completely, as the user needs also to add _distributeShardsLike to the collection marked as initial collection inside the graph definition.

`satellite`

    arangosh> db._createDocumentCollection("vertices2", { replicationFactor: "satellite" });

>     arangosh> db._createEdgeCollection("edges2", { replicationFactor: "satellite" });

Reviewer: That alone does not suffice; `distributeShardsLike` has to be set to the leading collection of the graph, which is the first vertex collection added to it.

Reviewer: Also, a line how these would be added to the actual graph would be good.

A full example
--------------

    arangosh> var explain = require("@arangodb/aql/explainer").explain
    arangosh> graphModule = require("@arangodb/general-graph");
    arangosh> satGraphModule = require("@arangodb/satellite-graph");
    arangosh> graphModule._create("normalGraph", [ graphModule._relation("edges", "vertices", "vertices") ], [], {});
    arangosh> satGraphModule._create("satelliteGraph", [ satGM._relation("satEdges", "satVertices", "satVertices") ], [], {});
    arangosh> db._create("collection", {numberOfShards: 8})
    arangosh> db._create("collection2", {numberOfShards: 8})

> Let's analyse a join involving a traversal.

Reviewer 2: ```suggestion
Let's analyse a join involving a traversal:
```

  2   EnumerateCollectionNode   DBS      0     - FOR doc IN collection   /* full collection scan, 8 shard(s) */
  8   RemoteNode                COOR     0       - REMOTE
  9   GatherNode                COOR     0       - GATHER 
  3   TraversalNode             COOR     0       - FOR v  /* vertex */, e  /* edge */, p  /* paths */ IN 1..1  /* min..maxPathDepth */ OUTBOUND 'vertices/start' /* startnode */  GRAPH 'normalGraph'
  4   CalculationNode           COOR     0         - LET #6 = [ doc, v, e, p ]   /* simple expression */   /* collections used: doc : collection */
  5   ReturnNode                COOR     0         - RETURN #6

Indexes used:
 By   Name   Type   Collection   Unique   Sparse   Selectivity   Fields        Ranges
  3   edge   edge   edges        false    false       100.00 %   [ `_from` ]   base OUTBOUND

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections  Options                                  Filter / Prune Conditions
 3   1..1   vertices            edges             uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   scatter-in-cluster
  2   remove-unnecessary-remote-scatter

> ```

Reviewer: I think it's possible to specify the commands and let the output be generated with the documentation. But its been a long time since I've used that last, so it might have changed. @Simran-B, can you shed some light on that?

Indexes used:
 By   Name   Type   Collection   Unique   Sparse   Selectivity   Fields        Ranges
 10   edge   edge   edges        false    false       100.00 %   [ `_from` ]   base OUTBOUND

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections  Options                                  Filter / Prune Conditions
 10  1..1                       edges             uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   scatter-in-cluster
  2   remove-unnecessary-remote-scatter
  3   scatter-satellite-graphs
  4   remove-satellite-joins
```

Note that now the `TraversalNode` (with id 10) is executed on each DBServer, leading to a great reduction in
required network communication, and hence potential gains in query performance.

Caveats
-------

The cluster will automatically distribute all satellite graphs to all DBServers, and apply
synchronous replication. 
This means that writes are executed on the leader only, and then have to be replicated
to all followers, leading to degradation in write performance.

There is also a chance that a write has not been replicated in time before a query is executed

> and a query involving satellite graphs returns out of date results.

Reviewer: If writes are done while a query is being instantiated, there is chance that a write has not been fully replicated. This can lead to inconsistent results.
```suggestion
If writes are done to a satellite graph while a query is being instantiated, there is chance that a write has not been fully replicated. This can lead to inconsistent results for queries involving that graph.
```

Author: Whats the difference between "instantiation" and "execution"?

I neither like "inconsistent" (because it doesn't say what it's inconsistent *with*), nor "out of date".

Reviewer: > Whats the difference between "instantiation" and "execution"?

Technically, after the planning stage, the query snippets get sent to all DBServers. There, they are instantiated, meaning an object is created in the query registry. Most importantly, the local transactions are opened, and thus RocksDB Snapshots are taken on each DBServer.

Only if writes happen while the transactions are started, it can happen that one DBServer gets a write before the transaction is opened, and another does not.

> I neither like "inconsistent" (because it doesn't say what it's inconsistent with), nor "out of date".

What I meant with inconsistent, and I agree that we should clarify that, is that the different replicas of the same satellite collection(s) have a different state (snapshot) at the very beginning of the query. Thus the same traversal will return different results on different DBServers. That can lead to all kind of breakage, depending on the query. We could even say this is undefined behaviour, as of course all optimizer rules assume this does not happen.
